File: backend/src/routes/alarms.js
const { addAudit, machineById, nowText } = require("../domain");
const { parseJsonBody, sendJson } = require("../http-utils");

async function handleAlarmRoutes(req, res, pathname, store) {
  const alarmAckMatch = pathname.match(/^\/api\/alarms\/([^/]+)\/ack$/);
  if (req.method === "POST" && alarmAckMatch) {
    const [, alarmId] = alarmAckMatch;
    const body = await parseJsonBody(req);
    const payload = await store.update((state) => {
      const alarm = state.alarms.find((item) => item.id === decodeURIComponent(alarmId));
      if (!alarm) {
        const error = new Error("Alarm not found");
        error.statusCode = 404;
        throw error;
      }
      alarm.status = "closed";
      alarm.acknowledgedAt = nowText();
      alarm.acknowledgedBy = body.actor || "Lab Operator";
      const machine = machineById(state, alarm.equipmentId);
      if (machine && machine.status === "alarm") machine.status = "maintenance";
      addAudit(state, `${alarm.id} 已確認處理`, alarm.acknowledgedBy);
      return { message: `${alarm.id} 已確認` };
    });
    sendJson(res, 200, payload);
    return true;
  }

  if (req.method === "POST" && pathname === "/api/alarms/simulate") {
    const body = await parseJsonBody(req);
    const payload = await store.update((state) => {
      const machine = state.equipment.find((item) => item.status !== "alarm") || state.equipment[0];
      machine.status = "alarm";
      const alarmId = `ALM-${String(state.alarmSeq).padStart(3, "0")}`;
      state.alarmSeq += 1;
      state.alarms.unshift({
        id: alarmId,
        equipmentId: machine.id,
        severity: "Medium",
        message: `${machine.name} 狀態異常，請確認機台訊號與保養紀錄。`,
        status: "alarm",
        createdAt: nowText()
      });
      addAudit(state, `${machine.name} 產生模擬告警 ${alarmId}`, body.actor || "System");
      return { message: `${alarmId} 已建立` };
    });
    sendJson(res, 201, payload);
    return true;
  }

  return false;
}

module.exports = {
  handleAlarmRoutes
};

File: backend/app/routes/alarms.py
from typing import Any
from urllib.parse import unquote

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..auth import require_roles
from ..http_utils import read_json_body
from ..services import alarm_service

router = APIRouter(prefix="/api/alarms", tags=["alarms"])


@router.get("")
async def list_alarms(request: Request) -> list[dict[str, Any]]:
    return request.app.state.store.read()["alarms"]


@router.post("/{alarm_id}/ack")
async def acknowledge_alarm(alarm_id: str, request: Request) -> dict[str, Any]:
    require_roles(request.state.user, "operator")
    body = await read_json_body(request)
    return alarm_service.acknowledge(
        request.app.state.store,
        alarm_id=unquote(alarm_id),
        actor=str(body.get("actor") or "Lab Operator"),
    )


@router.post("/simulate", response_model=None)
async def simulate_alarm(request: Request) -> JSONResponse:
    require_roles(request.state.user, "operator")
    body = await read_json_body(request)
    result = alarm_service.simulate(
        request.app.state.store,
        actor=str(body.get("actor") or "System"),
    )
    return JSONResponse(result, status_code=201)
